### PR TITLE
Make layout account for transforms

### LINKED ISSUE #528: Version of rotate that affects layout
Use case: Using rotated text in tables and grids.

As `rotate` does not change the elements bounds to its rotated bounding box, rotated text
in tables and grids is not easy to achieve.

```typst
#table(columns: (auto, auto),
  rotate([Test Text], -90deg),
  rotate([Test Text], 0deg))
```

Renders as:
grafik

But what I want is, that the first cell is auto-sized to the bounding box of the rotated element, not its original bounds. Every workaround I have tried failed, as I could not get the table cell to grow correctly.

Proposal: Either add a function that applies all transformations (e.g. `transform(rotate(...))`) and affects the layout or a parameter (or separater fn) for `rotate` to affect the layout

### PR DESCRIPTION
**Pull request by Dherse**
Closes #528

Allows the three transforms: `rotate` and `scale` to impact layout if the user sets the new optional (and settable) argument `layout` to `true`.

### CODE REVIEW

File: crates/typst-library/src/layout/transform.rs [hunk 1]
```diff
@@ -46,14 +46,14 @@ impl Layout for MoveElem {
         let delta = Axes::new(self.dx(styles), self.dy(styles)).resolve(styles);
         let delta = delta.zip_map(regions.base(), Rel::relative_to);
         frame.translate(delta.to_point());
-        Ok(Fragment::frame(frame))
+        return Ok(Fragment::frame(frame));
```
Comment (samueltardieu) [inline crates/typst-library/src/layout/transform.rs:49]: Why this change?

Comment (PgBiel) [reply crates/typst-library/src/layout/transform.rs:49]: Seems like a leftover from a previous attempt to make `move` impact layout as well

File: crates/typst-library/src/layout/transform.rs [hunk 2]
```diff
@@ -94,6 +94,20 @@ pub struct RotateElem {
     #[default(HAlign::Center + VAlign::Horizon)]
     pub origin: Align,
 
+    /// Whether the rotation impacts the layout.
+    ///
+    /// If set to `{false}`, the rotated content will be allowed to overlap
+    /// other content. However, when set to `{true}`, it will be compute the
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:100]:
```suggestion
    /// other content. However, when set to `{true}`, it will compute the
```

File: crates/typst-library/src/layout/transform.rs [hunk 3]
```diff
@@ -109,15 +123,24 @@ impl Layout for RotateElem {
...
+
+        // If we don't impact layout we exit early.
+        if !self.layout(styles) {
+            return Ok(Fragment::frame(frame));
+        }
+
+        // If we impact layout we wrap into a frame of the correct size
+        let (offset, size) = compute_bounding_box(&frame, ts);
+        let mut out = Frame::soft(size);
+        out.push(offset, FrameItem::Group(GroupItem::new(frame)));
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:142]: couldn't we transform the frame in place and avoid the extra nesting level? And even if not, `push_frame` is preferrable over creating a group item manually.

File: crates/typst/src/geom/transform.rs [hunk 4]
```diff
@@ -69,6 +69,13 @@ impl Transform {
         next.pre_concat(self)
     }
 
+    /// Transform a point.
+    pub fn transform_point(self, point: Point) -> Point {
+        let x = self.sx.of(point.x) + self.kx.of(point.y) + self.tx;
+        let y = self.ky.of(point.x) + self.sy.of(point.y) + self.ty;
+        Point { x, y }
+    }
```
Comment (laurmaedje) [inline crates/typst/src/geom/transform.rs:77]: this function already exists on `Point`

File: tests/typ/layout/transform-layout.typ [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// Test layout transformations
+---
+
+// Test that rotation impact layout.
+#set page(width: 200pt)
+#set rotate(layout: true)
+#let one(angle) = box(fill:blue, rotate([Test Text], angle * 1deg))
```
Comment (laurmaedje) [inline tests/typ/layout/transform-layout.typ:7]:
```suggestion
#let one(angle) = box(fill: aqua, rotate(angle)[Test Text])
```

File: tests/typ/layout/transform-layout.typ [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+#for angle in range(0, 360, step: 15) {
+  one(angle)
```
Comment (laurmaedje) [inline tests/typ/layout/transform-layout.typ:10]:
```suggestion
  one(angle * 1deg)
```

File: tests/typ/layout/transform-layout.typ [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+// Test that scaling impact layout.
+#set page(width: 200pt)
+#set text(size: 32pt)
+#let scaled(body) = box(scale(x: 20%, y: 40%, body))
+
+#set scale(layout: false)
+Hello #scaled[World]!
+
+
```
Comment (laurmaedje) [inline tests/typ/layout/transform-layout.typ:23]:
```suggestion
```

**Review by laurmaedje — COMMENTED** [review]
As said on Discord, I think this behaviour should be the new default. 

I also wonder whether `layout` is the best name for this. Here are a few alternatives @reknih and I came up with:

- `bounds: "original" | "transformed"`
- `reflow: false | true`
- `resize: false | true`
- `affect: false | true`

**Comment by PgBiel** [comment]
> As said on Discord, I think this behaviour should be the new default.
> 
> I also wonder whether `layout` is the best name for this. Here are a few alternatives @reknih and I came up with:
> 
>     * `bounds: "original" | "transformed"`
> 
>     * `reflow: false | true`
> 
>     * `resize: false | true`
> 
>     * `affect: false | true`

I think `affect` isn't too specific regarding what it is affecting, and `resize` may sound redundant since `scale` does resize either way (just in a different manner depending on whether layout is affected or not), so the two first alternatives are better imo.

Of the two, I'd personally prefer `reflow`, but I'd be open to more suggestions as well.

File: crates/typst-library/src/layout/transform.rs (hunk 1 above)
Comment (Dherse) [reply crates/typst-library/src/layout/transform.rs:49]: @PgBiel you are absolutely correct, thank you two for pointing it out <3

File: crates/typst-library/src/layout/transform.rs [hunk 8]
```diff
@@ -53,7 +53,7 @@ impl Layout for MoveElem {
 /// Rotates content without affecting layout.
 ///
 /// Rotates an element by a given angle. The layout will act as if the element
-/// was not rotated.
+/// was not rotated unless you specify `{layreflowout: true}`.
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:56]:
```suggestion
/// was not rotated unless you specify `{reflow: true}`.
```

File: crates/typst-library/src/layout/transform.rs (hunk 3 above)
Comment (laurmaedje) [reply crates/typst-library/src/layout/transform.rs:168]: This was not addressed.

File: crates/typst-library/src/layout/transform.rs [hunk 9]
```diff
@@ -94,6 +94,24 @@ pub struct RotateElem {
     #[default(HAlign::Center + VAlign::Horizon)]
     pub origin: Align,
 
+    /// Whether the rotation impacts the layout.
+    ///
+    /// If set to `{false}`, the rotated content will be allowed to overlap
+    /// other content. However, when set to `{true}`, it will compute the
+    /// new size of the rotated content and adjust the layout accordingly.
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:101]:
```suggestion
    /// If set to `{false}`, the rotated content will retain the bounding box of
    /// the original content. If set to `{true}`, the bounding box will take the
    /// rotation of the content into account and adjust the layout accordingly.
```
same for scale

File: crates/typst-library/src/layout/transform.rs [hunk 10]
```diff
@@ -94,6 +94,24 @@ pub struct RotateElem {
...
+    ///
+    /// ```example
+    /// #let rotated(body) = rotate(90deg, reflow: true, body)
+    ///
+    /// Hello #rotated[World]!
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:106]:
```suggestion
    /// Hello #rotate(90deg, reflow: true)[World]!
```
same for scale

File: crates/typst-library/src/layout/transform.rs [hunk 11]
```diff
@@ -94,6 +94,24 @@ pub struct RotateElem {
...
+    /// ```
+    ///
+    /// Using this is equivalent to using the [`measure`]($measure) function
+    /// to measure the size of the content and then wrapping this element in
+    /// a box with the computed size rotated by `angle`.
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:111]: I'm not sure this paragraph is really useful, but I might be wrong.

File: crates/typst-library/src/layout/transform.rs [hunk 12]
```diff
@@ -172,16 +239,71 @@ impl Layout for ScaleElem {
...
         let mut frame = self.body().layout(vt, styles, pod)?.into_frame();
-        let Axes { x, y } = self
-            .origin(styles)
-            .resolve(styles)
-            .zip_map(frame.size(), FixedAlign::position);
-        let transform = Transform::translate(x, y)
-            .pre_concat(Transform::scale(self.x(styles), self.y(styles)))
+        let Axes { x, y } = align.zip_map(frame.size(), FixedAlign::position);
+
+        // Apply the transform.
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/transform.rs:277]: I feel like there is quite a lot of code duplication in the two code paths. Would it be possibly to unify them?

File: crates/typst/src/geom/point.rs [hunk 13]
```diff
@@ -51,13 +51,25 @@ impl Point {
     }
 
     /// Transform the point with the given transformation.
+    ///
+    /// In the event that one of the coordinates is infinite,
```
Comment (laurmaedje) [inline crates/typst/src/geom/point.rs:55]: Please wrap comments at 80 columns, not earlier.

File: tests/typ/layout/transform-layout.typ [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+#for angle in range(0, 360, step: 15) {
+  one(angle * 1deg)
+}
+
+---
+
+// Test relative sizing in rotated boxes.
+#set page(width: 200pt, height: 200pt)
+#set text(size: 32pt)
+#let rotated(body) = box(rotate(90deg, box(stroke: 0.5pt, height: 20%, clip: true, body)))
```
Comment (laurmaedje) [inline tests/typ/layout/transform-layout.typ:18]: Please add a linebreak somehwere here and at the long line below.

File: tests/typ/layout/transform-layout.typ [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+#set page(width: 200pt)
+#set rotate(reflow: true)
+
+#let one(angle) = box(fill: aqua, rotate(angle)[Test Text])
+#for angle in range(0, 360, step: 15) {
+  one(angle * 1deg)
+}
+
+---
+
```
Comment (laurmaedje) [inline tests/typ/layout/transform-layout.typ:14]:
```suggestion
```
No need for a blank line between `---` and the comment.

**Comment by laurmaedje** [comment]
As mentioned on Discord, there are still some problems here. Can those be worked out or do you think this should be closed and postponed to a future layout engine rework?

**Comment by PgBiel** [comment]
Are there any blockers besides the reviews? This would definitely be interesting to see implemented.

**Comment by laurmaedje** [comment]
Thank you!
